compensations/src/test/java/org/jboss/narayana/compensations/functional/transaction/TransactionLocalTest.java

```diff
@@ -0,0 +1,128 @@
+import org.jboss.arquillian.junit.Arquillian;
+import org.jboss.narayana.compensations.api.CompensationHandler;
+import org.jboss.narayana.compensations.api.ConfirmationHandler;
+import org.jboss.narayana.compensations.api.Transaction;
+import org.jboss.narayana.compensations.functional.common.DummyData;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import javax.inject.Inject;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+/**
+ * @author <a href="mailto:gytis@redhat.com">Gytis Trikleris</a>
+ */
+@RunWith(Arquillian.class)
+public class TransactionLocalTest {
+
+    @Inject
+    private Transaction transaction;
+
+    @Deployment
+    public static WebArchive createTestArchive() {
+        WebArchive archive = ShrinkWrap.create(WebArchive.class, "test.war")
+                .addClass(DummyData.class) // Deployment need at least one class with Compensations API annotations to get dependencies
+                .addAsWebInfResource(EmptyAsset.INSTANCE, "beans.xml");
+
+        return archive;
+    }
+
+    @Test
+    public void successfulScenario() {
+        final Counter actionsCounter = new Counter();
+        final Counter compensationsCounter = new Counter();
+        final Counter confirmationsCounter = new Counter();
+
+        transaction
+                .doAction(actionsCounter::increment)
+                .doAction(actionsCounter::increment,
+                        compensationsCounter::increment,
+                        confirmationsCounter::increment)
+                .doAction(actionsCounter::increment,
+                        (CompensationHandler) compensationsCounter::increment)
+                .doAction(actionsCounter::increment,
+                        (ConfirmationHandler) confirmationsCounter::increment)
+                .execute();
+
+        assertEquals(4, actionsCounter.get());
+        assertEquals(2, confirmationsCounter.get());
+        assertEquals(0, compensationsCounter.get());
+    }
+
+    @Test
+    public void exceptionScenario() {
+        final Counter actionsCounter = new Counter();
+        final Counter compensationsCounter = new Counter();
+        final Counter confirmationsCounter = new Counter();
+
+        try {
+            transaction
+                    .doAction(actionsCounter::increment)
+                    .doAction(actionsCounter::increment,
+                            compensationsCounter::increment,
+                            confirmationsCounter::increment)
+                    .doAction(actionsCounter::increment,
+                            (CompensationHandler) compensationsCounter::increment)
+                    .doAction(actionsCounter::increment,
+                            (ConfirmationHandler) confirmationsCounter::increment)
```

Comment:
Here example from above with implied interface implementation does not work. Unspecified implementation is ambiguous for the compiler, because there are two Transaction.doAction methods which take two arguments. Therefore, we need to specifically cast lambda to tell compiler which actual method are we using.